CHANGELOG.md

```diff
@@ -2,6 +2,22 @@
 
 ## Upcoming version
 
+### Added
+
+- \[[#307](https://github.com/rust-vmm/vm-memory/pull/304)\] Introduce `VolatileSlice::truncate`
+
+### Changed
+
+- \[[#307](https://github.com/rust-vmm/vm-memory/pull/304)\] Move `read_volatile_from`, `read_exact_volatile_from`,
+  `write_volatile_to` and `write_all_volatile_to` functions from the `GuestMemory` trait to the `Bytes` trait.
+
+### Removed
+
+- \[[#307](https://github.com/rust-vmm/vm-memory/pull/304)\] Remove deprecated functions `Bytes::read_from`, `Bytes::read_exact_from`,
+  `Bytes::write_to`, `Bytes::write_all_to`, `GuestMemory::as_slice`, `GuestMemory::as_slice_mut`, `GuestMemory::with_regions`, 
```

Comment:
I agree that the non-volatile read/write method are not sound for `Bytes`, but maybe `read_exact_from`/`write_all_to` could be added to `ByteValued`?

Reply:
Yeah, that makes sense to me! There's some longer-term plans to eliminate `ByteValued` and switch to using zerocopy (https://github.com/rust-vmm/vm-memory/issues/246), which provides the same functionality. Their traits do actually implement the `read_exact_from`/`write_all_to` functionality you're referring to for free (e.g. https://docs.rs/zerocopy/latest/zerocopy/trait.IntoBytes.html#method.write_to_io).

Reply:
Yes, it makes sense to make `ByteValued` a thin wrapper over zerocopy, wrapping FromBytes/IntoBytes/AsBytes and providing the implementation of the legacy vm-memory API.

But even if it's going to be a legacy API it will take time to phase it out from everywhere, so let's add the two to ByteValued anyway.

Reply:
Mhh, the possibility of turning it into a wrapper for a phase-out period is certainly enticing... maybe some `zerocopy` feature that if enabled provides a blanket impl for ByteValued in terms of zerocopy, and causes rustc to warn on manual impls of the trait (if that's possible), and then later the default impl can be promoted to trait bounds, and even later to a removal of ByteValued altogether once everyone only uses zerocopy types anymore. I'll have to think about that, thanks!

For now, I've added the read/write variants as suggested, since they won't cause any harm indeed :) 
(and also fixed up two minor things I noticed about ByteValued while I was at it)

Reply:
Since `read_exact_from`/`write_all_to` were already deprecated, I would not move them to another trait and keep them deleted. Looking at the impls: `write_all_to` is just indirection one liner, and `read_exact_from` is a constructor with incorrect name. Do we really need to keep them around?

Reply:
They were deprecated because they're unsound in Bytes context, but they're useful nevertheless and make it easier to write correct code (for example avoiding that you use read/write instead of read_exact/write_all). The PR just moves them to the right place.